pinot-segment-local/src/main/java/org/apache/pinot/segment/local/segment/index/readers/forward/VarByteChunkSVForwardIndexReader.java

```diff
@@ -85,6 +85,10 @@ private String getStringCompressed(int docId, ChunkReaderContext context) {
 
     int length = valueEndOffset - valueStartOffset;
     byte[] bytes = _reusableBytes.get();
+    if (bytes.length < _lengthOfLongestEntry) {
```

Comment:
When do we reduce the byte[] length? For instance, say we onboard a table which has very long entries. Running some queries will set the length of the byte[] to a large value for all threads. Even if a user deletes their table this excess memory won't be reclaimed.

Reply:
we will not reduce the byte length. Since it's a static field, so each thread will only have one copy, I think it's acceptable.

Reply:
The V4 writer allocate byte[] in every function call to avoid this problem. I am also consider about moving the byte[] to ReaderContext class so that each Column per query can share the same byte[]. And after query is finished, the ReaderContext will be closed.

Reply:
This PR is a nice improvement. But I think it would be better to do what you said here: To add the bytes[] to the context.

Reply:
*static threadLocal byte[] of IndexReader*
Pros: allocated only once per segment per thread;
Cons: memory won't be deallocated even no queries.

*threadLocal byte[] of ReaderContext and implement close() method to remove threadLocal*
Pros: deallocate the memory in time when query gets finished.
Cons: repetitive allocation of byte[] array per query, query latency might tumble; need to carefully check the usage of ReaderContext to ensure the close() will be invoked finally after the query is completed (either failure or success)

Reply:
>  repetitive allocation of byte[] array per query, query latency might tumble

This is just a single byte[] that is taken for few ms, so it should never reach the old gen.

> need to check the usage of ReaderContext carefully to ensure the close() will be invoked finally after the query is completed

That is not like that AFAIU. First, the context should be closed right now. Otherwise other indexes may be leaking. Even if there is a bug and it is not being closed, what should be clear is that we should not have a reference to that context once the query finishes, so it will be eventually (and again, probably in the young gen) get collected, which will collect the byte[].

Reply:
Just to be clear: The original reason to have a ThreadLocal was to do not allocate one byte[] per row, which is expensive. Allocating one byte[] per segment is totally fine. We allocate quite more than that in other parts of the same path. And usually it is a bad thing to be too aggressive at catching in a managed environment like the JVM

Reply:
Do you mean allocating one byte[] per Context is totally fine?

Cache byte[] per query worker is not aggressive as the screenshot shows.

Reply:
It is clear that having a `ThreadLocal<byte[]>` per column is bad. My understanding is that you are proposing two solutions:
1. Either use a single ThreadLocal byte[] per thread.
2. Move the byte[] to reader context.

I think `2` is better, but I'm fine with `1` as well.

Reply:
I will test approach 2. the approach 1's result is already attached in the description.